src/EditorFeatures/Core/Implementation/CodeDefinitionWindow/DefinitionContextTracker.cs

```diff
@@ -0,0 +1,240 @@
+            if (pointInRoslynSnapshot == null)
+            {
+                return;
+            }
+
+            _currentUpdateCancellationToken = new CancellationTokenSource();
+
+            var cancellationToken = _currentUpdateCancellationToken.Token;
+            var asyncToken = _asyncListener.BeginAsyncOperation(nameof(DefinitionContextTracker) + "." + nameof(QueueUpdateForCaretPosition));
+            UpdateForCaretPositionAsync(pointInRoslynSnapshot.Value, cancellationToken).CompletesAsyncOperation(asyncToken);
+        }
+
+        private async Task UpdateForCaretPositionAsync(SnapshotPoint pointInRoslynSnapshot, CancellationToken cancellationToken)
+        {
+            var locations = await GetContextFromPointAfterDelayAsync(pointInRoslynSnapshot, cancellationToken).ConfigureAwait(true);
+
+            await _codeDefinitionWindowService.SetContextAsync(locations, cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task<ImmutableArray<CodeDefinitionWindowLocation>> GetContextFromPointAfterDelayAsync(
+            SnapshotPoint pointInRoslynSnapshot, CancellationToken cancellationToken)
+        {
+            // TODO: Does this allocate too many tasks - should we switch to a queue like the classifier uses?
+            await Task.Delay(TimeSpan.FromMilliseconds(250), cancellationToken).ConfigureAwait(false);
+
+            var document = pointInRoslynSnapshot.Snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                return ImmutableArray<CodeDefinitionWindowLocation>.Empty;
+            }
+
+            return await GetContextFromPointAsync(document, pointInRoslynSnapshot.Position, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Internal for testing purposes.
+        /// </summary>
+        internal async Task<ImmutableArray<CodeDefinitionWindowLocation>> GetContextFromPointAsync(
+            Document document, int position, CancellationToken cancellationToken)
+        {
+            var workspace = document.Project.Solution.Workspace;
+            if (!document.SupportsSemanticModel)
+            {
+                var goToDefinitionService = document.GetRequiredLanguageService<IGoToDefinitionService>();
+                var navigableItems = await goToDefinitionService.FindDefinitionsAsync(document, position, cancellationToken).ConfigureAwait(false);
+                if (navigableItems != null)
+                {
+                    var navigationService = workspace.Services.GetService<IDocumentNavigationService>();
+
+                    var builder = new ArrayBuilder<CodeDefinitionWindowLocation>();
+                    foreach (var item in navigableItems)
+                    {
+                        if (navigationService.CanNavigateToPosition(workspace, item.Document.Id, item.SourceSpan.Start, cancellationToken))
```

Comment:
i believe this needs to be on the UI thread.  alternatively use (or add) an overload of this that is async.

Reply:
Now using the async version.